# Homework 5
Zach Petko

3/2/2025

In [2]:
# import neccesary libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score


### Part 1: Data Loading and Exploration

In [3]:
# 1. Load the California Housing dataset from sklearn.datasets.
from sklearn.datasets import fetch_california_housing

# Load the housing dataset
housing = fetch_california_housing()



In [4]:
# 2. Create a Pandas DataFrame for the features and a Series for the target variable (med_house_value). 
X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

In [5]:
# 3. Perform an initial exploration of the dataset:

## Display the first five rows of the dataset
print("First 5 rows of X: ")
display(X.head())
print("First 5 rows of y: ")
display(y.head())

## Print the feature names and check for missing values
print(f"Column Names: {X.columns.tolist()} \n")
print("Number of NA values in each column of X: ")
display(X.isna().sum())

## Generate summary statistics (mean, min, max, etc.)
print("Summary Statistics for X: ")
display(X.describe())


First 5 rows of X: 


MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

First 5 rows of y: 


0    4.526
1    3.585
2    3.521
3    3.413
4    3.422
Name: med_house_value, dtype: float64

Column Names: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude'] 

Number of NA values in each column of X: 


MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64

Summary Statistics for X: 


MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude  
count  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704  
std       10.386050      2.135952      2.003532  
min        0.692308     32.540000   -124.350000  
25%        2.429741     33.930000   -121.800000  
50%        2.818116     34.260000   -118.490000  
75%        3.282261     37.710000   -118.010000  
max     1243.333333     41.950000   -114.310000

## Part 2: Linear Regression on Unscaled Data

In [6]:
# 4. Split the dataset into training and test sets (80% training, 20% testing).

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [7]:
# 5. Train a linear regression model on the unscaled data using

lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)
print("Linear Regression Summary: ")
lin_reg_summary = pd.DataFrame({
    "Variable": ['Intercept', 'MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 
                 'Population', 'AveOccup', 'Latitude', 'Longitude'],
    "Coefficent": [lin_reg.intercept_] + list(lin_reg.coef_)
})
display(lin_reg_summary)


Linear Regression Summary: 


Variable  Coefficent
0   Intercept  -37.023278
1      MedInc    0.448675
2    HouseAge    0.009724
3    AveRooms   -0.123323
4   AveBedrms    0.783145
5  Population   -0.000002
6    AveOccup   -0.003526
7    Latitude   -0.419792
8   Longitude   -0.433708

In [8]:
# 6. Make predictions on the test set

y_pred = lin_reg.predict(X_test)
print("Predictions of y using Linear Regression: ")
print(y_pred)

Predictions of y using Linear Regression: 
[0.71912284 1.76401657 2.70965883 ... 4.46877017 1.18751119 2.00940251]


In [9]:
# 7. Evaluate model performance using...

# Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred)
print("Unscaled Data Model:")
print(f"Mean Squared Error: {mse:.2f}")

# Root Mean Squared Error (RMSE)
rmse = root_mean_squared_error(y_test, y_pred)
print(f"Root Squared Error: {rmse:.2f}")

# R^2 Score
r2 = r2_score(y_test, y_pred)
print(f"R² Score: {r2:.2f}")

Unscaled Data Model:
Mean Squared Error: 0.56
Root Squared Error: 0.75
R² Score: 0.58


8. Interpretation Questions:

What does the R² score tell us about model performance?

    - The R² value of 0.58 means that 58% of the variation in y is explained by using the variables in X used in the linear regression model. 

Which features seem to have the strongest impact on predictions based on the model’s coefficients?

    - The 3 variables with the largest impact on the Median Housing Value is in descending order AveBedrms(0.783145), MedInc(0.448675), and Longitude(-0.433708). This is beacuse these are the three variables with the highest estimated coeficents in abosulte value. This means that a 1 unit change in these variables will have the highest change in the y variable (Median Housing Value).

    - AveBedrms(average bedrooms) has the largest coeficent in aboslute value and is positive. This means that AveBedrms has the largest 1 unit change effect on Median Housing Value. So, 1 additional average bedroom predicts a 0.783145 increase in median housing value.

     - MedInc(median income) is positive. So, 1 additional unit of median income predicts a 0.448675 increase in median housing value.

     - Longitude is negative. So, 1 additional unit of longitude predicts a 0.433708 decrease in median housing value.


How well do the predicted values match the actual values?

    - The MSE value is 0.56. This means that the average squared error between the prdicted and actual values is 0.56. A lower value is better, but the RMSE is easier to interpret. The RMSE value is 0.75. This means that on average the predicted value from our model is off by 0.75 units from the actual value. When considering the values of y, our model does a reasonably good job of predicting the actual values.

    - As discussed above, the R² means that our model using the independent X variables explains about 58% of the variation in y. This is not extraordinarily high but is reasonable considering the area of study. Overall the model does a mmoderately good job of predicting the actual values.

## Part 3: Scaled Model

In [10]:
from sklearn.preprocessing import StandardScaler

# Initialize the scaler and apply it to the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled = pd.DataFrame(X_scaled, columns=X.columns)
X_scaled

# Split Scaled Data
X_train_scaled,X_test_scaled, y_train_scaled, y_test_scaled =train_test_split(X_scaled,y,test_size=0.2, random_state=42)

# Fit the scaled data
lin_reg_scaled = LinearRegression()
lin_reg_scaled.fit(X_train_scaled, y_train_scaled)

# Make Predictions
y_pred_scaled = lin_reg_scaled.predict(X_test_scaled)
y_pred_scaled

# Evaluate model performance
mse = mean_squared_error(y_test_scaled, y_pred_scaled)
rmse = root_mean_squared_error(y_test_scaled, y_pred_scaled)
r2 = r2_score(y_test_scaled, y_pred_scaled)

print("Scaled Data Model:")
print(f"Mean Squared Error: {mse:.2f}")
print(f"Root Squared Error: {rmse:.2f}")
print(f"R² Score: {r2:.2f}")
print("Model Coefficients (Scaled):")
print(pd.Series(lin_reg_scaled.coef_, index=X.columns))

Scaled Data Model:
Mean Squared Error: 0.56
Root Squared Error: 0.75
R² Score: 0.58
Model Coefficients (Scaled):
MedInc        0.852382
HouseAge      0.122382
AveRooms     -0.305116
AveBedrms     0.371132
Population   -0.002298
AveOccup     -0.036624
Latitude     -0.896635
Longitude    -0.868927
dtype: float64


## Part 4: Feature Selection and Simplified Model

 13. Select three features from the dataset to build a simplified model. Explain your choice. 

    - For the simplified model, I will choose the following variables: AveBedrms, MedInc, and AveRooms. This is because they have the highest coefficent values in absolute values (apart for latetude and longitude) and they seem to have the most intuitive relationship with housing price. Although the latetude and longitude have higher coefficent values, I chose to avoid them due to the difficulty in interpreting these variables. 

In [11]:
# 14. Train a new linear regression model using only these three features.

#select relevent columns
X_train2 = X_train[['AveBedrms', 'MedInc', 'AveRooms']] 
X_test2 = X_test[['AveBedrms', 'MedInc', 'AveRooms']]

#run regression
lin_reg2 = LinearRegression()
lin_reg2.fit(X_train2, y_train)
print("Linear Regression Summary: ")
lin_reg2_summary = pd.DataFrame({
    "Variable": ['Intercept', 'AveBedrms', 'MedInc', 'AveRooms'],
    "Coefficent": [lin_reg2.intercept_] + list(lin_reg2.coef_)
})
display(lin_reg2_summary)

Linear Regression Summary: 


Variable  Coefficent
0  Intercept    0.021819
1  AveBedrms    1.170217
2     MedInc    0.542349
3   AveRooms   -0.246163

In [12]:
# 15. Evaluate the performance of this simplified model and compare it to the full model

#create predictions
y_pred2 = lin_reg2.predict(X_test2)
print("Predictions of y using Simplified Linear Regression: ")
print(y_pred2)

#get performance stats

# Mean Squared Error (MSE)
mse2 = mean_squared_error(y_test, y_pred2)

# Root Mean Squared Error (RMSE)
rmse2 = root_mean_squared_error(y_test, y_pred2)

# R^2 Score
r22 = r2_score(y_test, y_pred2)

#Create comparison table
perf_summary = pd.DataFrame({
    "Variable": ['MSE', 'RMSE', 'R^2'],
    "Big Model": [mse,rmse,r2],
    "Simplified Model": [mse2,rmse2,r22]
})
print("\nSummary Statistics Comparison:")
display(perf_summary)

Predictions of y using Simplified Linear Regression: 
[1.09794478 1.55080224 2.31795302 ... 4.35435412 1.38076318 2.17579031]

Summary Statistics Comparison:


Variable  Big Model  Simplified Model
0      MSE   0.555892          0.675854
1     RMSE   0.745581          0.822104
2      R^2   0.575788          0.484242

We see that the simplified model is worse in all of the considered statistics (MSE, RMSE, and R²). This means that although this model is smaller, it does not fit the testing data as well as the larger model with more independent variables. This is because the MSE and RMSE is larger in the simplified model than the big model, meaning that on average the error is larger in the simplified model. Additionally, the lower R² in the simplified model means that the simplified model explains less variance than the larger model. Therefore the big model has much better performance than the simplified model.

16. Interpretation Questions:

How does the simplified model compare to the full model?

    - As discussed above, the simplified has less coeficents and is smaller but is less powerful as a predictive model since it performs much worse in on the testing dataset. The simplified model is more efficent since there are less calculations required to estimate the model and then less to estimate y. However, linear regression is not very computationally intensive so this should not make a huge difference in reality.

Would you use this simplified model in practice? Why or why not?

    - I would not use the simplified model soley based on its performance. For a linear regression model, it does not perform well enough on the training data and its errors are far to large to be effecive in practice. I think with some modifications there could be a good model created, but I would not feel comfortable using this one in practice.
